### ISSUE TITLE
Configurable reconnects

### ISSUE THREAD

**Issue by rokups** [n1nj4sec/pupy#426]
Could we have configurable reconnect parameters? I suggest following:

* Reconnect delay, set 0 for no delay.
* Random deviation of reconnect timeout, set 0 for no deviation.
* Number of reconnect retries, set 0 for no limit.

At the moment if i create say `ps1_oneliner` payload, execute it and it connects, if connection is lost it will keep retrying connection without sleeping. Naturally this means lots of noise on the network which makes pupy easy to notice.

If there were said parameters it would be possible to either keep current behavior or to configure pupy so that reconnects are few and timing is not consistent. After set number of failed retries pypy should exit.

**Comment by rokups** [comment]
I started looking at the code more closely and i see pupy has variable reconnect times, though it would still be good to make that part more configurable. I started coding and this is what i got so far is passing a string command line parameter to `pp.y` which describes reconnect pattern.

`./pp.py --reconnect-pattern="120, 0.5, 1;  320, 3, 5;  15, 30"` - defines current reconnect scheme. Up to 120 attempts sleep between 0.5s and 1s. Up to 320 reconnects sleep between 3s and 5s. Above 320 reconnects sleep between 15s and 30s.

Pattern structure should be pretty obvious here: "<max_attempt>, <min_seconds>, <max_seconds>; [...]", If `max_attempt` is not specified then it is a sleep range for all remaining tries. Not specifying sleep times for all remaining tries makes `./pp.py` exit. I hope it is not too complex? 

`./pp.py --reconnect-pattern="3, 10, 30"` would try reconnecting up to 3 times and exit.

Please check out [the code](https://github.com/rokups/pupy/commit/bd0d08e01df951a55ff87278496aeaea9b1db6ea). This is not a complete solution. Is this an acceptable approach? Should i work on it further? Have any comments on how it should be done instead?

**Comment by n1nj4sec** [comment]
Hello ! 
Sry I may have missed your message, yes it looks nice but I don't see the problem with the current reconnect pattern and a configurable reconnect pattern seems overkill :)

**Comment by rokups** [comment]
Fair enough ;) My proposal was rather complex, i implemented it as `--reconnect-delay 120 0.5 1 --reconnect-default 15 30` which translates to "sleep between 0.5 and 1s when try count is less or equal to 120". After 120 tries it would sleep between 15s and 30s. Multiple `--reconnect-delay` parameters may be used. Maybe this is bit more proper and less overkill? ;)
